model/map/src/main/java/org/keycloak/models/map/userSession/MapUserSessionProvider.java

```diff
@@ -0,0 +1,777 @@
+        userSessionTx.updateIfChanged(origEntity.getId(), res, MapUserSessionEntity::isUpdated);
+        return res;
+    }
+
+    private MapAuthenticatedClientSessionEntity registerEntityForChanges(MapAuthenticatedClientSessionEntity origEntity) {
+        MapAuthenticatedClientSessionEntity res = clientSessionTx.read(origEntity.getId(), id -> Serialization.from(origEntity));
+        clientSessionTx.updateIfChanged(origEntity.getId(), res, MapAuthenticatedClientSessionEntity::isUpdated);
+        return res;
+    }
+
+    private MapUserLoginFailureEntity registerEntityForChanges(MapUserLoginFailureEntity origEntity) {
+        MapUserLoginFailureEntity res = userLoginFailureTx.read(origEntity.getId(), id -> Serialization.from(origEntity));
+        userLoginFailureTx.updateIfChanged(origEntity.getId(), res, MapUserLoginFailureEntity::isUpdated);
+        return res;
+    }
+
+    @Override
+    public AuthenticatedClientSessionModel createClientSession(RealmModel realm, ClientModel client, UserSessionModel userSession) {
+        MapAuthenticatedClientSessionEntity entity =
+                new MapAuthenticatedClientSessionEntity(UUID.randomUUID(), realm.getId(), client.getId(), false);
+
+        LOG.tracef("createClientSession(%s, %s, %s)%s", realm, client, userSession, getShortStackTrace());
+
+        clientSessionTx.create(entity.getId(), entity);
+
+        MapUserSessionEntity userSessionEntity = getUserSessionById(UUID.fromString(userSession.getId()));
+
+        if (userSessionEntity == null) {
+            throw new IllegalStateException("User session entity does not exist: " + userSession.getId());
+        }
+
+        userSessionEntity.addAuthenticatedClientSession(client.getId(), entity.getId());
+
+        return clientEntityToAdapterFunc(realm, client, userSession).apply(entity);
+    }
+
+    @Override
+    public AuthenticatedClientSessionModel getClientSession(UserSessionModel userSession, ClientModel client,
+                                                            UUID clientSessionId, boolean offline) {
+        LOG.tracef("getClientSession(%s, %s, %s, %s)%s", userSession, client,
+                clientSessionId, offline, getShortStackTrace());
+
+        Objects.requireNonNull(userSession, "The provided user session cannot be null!");
+        Objects.requireNonNull(client, "The provided client cannot be null!");
+        if (clientSessionId == null) {
+            return null;
+        }
+
+        ModelCriteriaBuilder<AuthenticatedClientSessionModel> mcb = clientSessionStore.createCriteriaBuilder()
+                .compare(AuthenticatedClientSessionModel.SearchableFields.ID, ModelCriteriaBuilder.Operator.EQ, clientSessionId)
+                .compare(AuthenticatedClientSessionModel.SearchableFields.CLIENT_ID, ModelCriteriaBuilder.Operator.EQ, client.getId())
+                .compare(AuthenticatedClientSessionModel.SearchableFields.IS_OFFLINE, ModelCriteriaBuilder.Operator.EQ, offline);
```

Comment:
userSession is missing and should be also added to restrict chance for collision attacks. Also realm_id should be included in the criteria